### ISSUE TITLE
IMAP doesn't sync correctly - BAD 'Command Error' following APPEND

### ISSUE THREAD

**Issue by HenkDemper** [thunderbird/thunderbird-android#1809]
### Expected behavior
After setting up an IMAP account in K9 mail I'm having perfect 2-way synchronisation with a Desktop mail client application. Everything I do (read, delete) in either application reflects in the other. I would have expected that it stays this way.

### Actual behavior
After using it for some month perfectly this way, suddenly about 2 weeks ago I got syncing errors and they do not go away. Issues observed:

1. When I read a mail in K9, it is marked as such. But after the next sync, it is marked as unread again. Older messages from before the issue (older than 2 weeks when it started happening) stay marked as read (as they should). It seems the 'read' mail status is not propagated.

2. When I delete a mail in K9, I do not see it anymore in the list, even after a new sync as expected. But if then goto the Desktop mail client, the deleted messages still appear. It seems the deleted messages in K9 are not deleted from the Server. This used to work perfectly.

### Steps to reproduce

For the 'read mail' issue:

1. Read an unread mail in K9 -> it is marked as such
2. Sync mail again -> the mail is marked as unread again
3. Steps 1 and 2 can be repeated...

For the 'delete mail' issue:

1. Delete a mail from K9 -> it is removed from the mail list
2. Sync mail again -> it does not re-appear again (good)
3. Sync mail in Desktop mail client -> deleted mail still appears

### Environment
K-9 Mail version: Latest 5.010

Android version: 6.0.1 on Rooted Samsung GT-N7100

Account type (IMAP, POP3, WebDAV/Exchange): IMAP

### Attached Logs

I have attached 5 logs (filtered from logcat using 'k9' Regex) in chronological order

1. Read Mail.txt -> Read 1 mail on account 'Henk @ Primestream'
2. Sync Mail.txt -> Sync mail directly after that (log might show traces of multiple IMAP account)
3. Sync Mail (Warning Level).txt -> Same as above, but filtered for 'Warning Level' messages
4. Delete Mail.txt -> Delete one mail from account 'Henk @ Primestream'
5. Sync Mail.txt -> Sync mail directly after that

[1. Read Mail.txt](https://github.com/k9mail/k-9/files/605836/1.Read.Mail.txt)
[2. Sync Mail.txt](https://github.com/k9mail/k-9/files/605833/2.Sync.Mail.txt)
[3. Sync Mail (Warning Level).txt](https://github.com/k9mail/k-9/files/605835/3.Sync.Mail.Warning.Level.txt)
[4. Delete Mail.txt](https://github.com/k9mail/k-9/files/605837/4.Delete.Mail.txt)
[5. Sync Mail.txt](https://github.com/k9mail/k-9/files/605834/5.Sync.Mail.txt)

### Errors I see in logs

Looking at the logs, I see at least the following when Syncing mail:

```
com.fsck.k9.mail.MessagingException: Could not find UID for message based on Message-ID
Caused by: java.io.IOException: readStringUntil(): end of stream reached
```

Might it be that the issues start happening when there are too many items in a mail folder: I was experiencing the problems during a holiday, when the number of items in the Inbox folder exceeded a few hundred (now it is around 300). Normally I keep the Inbox 'clean' (around 50 messages or so).

**Comment by philipwhiuk** [comment]
Interesting. Just before the exception following the append we get  [BAD, Command Error. 10] but with an unexpected tag.

The series of messages is odd - a continuation followed by a BAD command with an unknown tag.

Because the tag doesn't match the message we sent we ignore it.

But it looks like from that point onwards the connection is dead because we then don't get a response to our UID SEARCH (hence end of stream).

I'm not sure why outlook is throwing a fit here.

Any ideas @cketti ?

```
11-22 09:19:32.632 12496-12514/com.fsck.k9 V/k9: conn42368062>>> 5 NOOP
11-22 09:19:32.779 12496-12514/com.fsck.k9 V/k9: conn42368062<<<#5# [OK, NOOP completed.]
11-22 09:19:32.788 12496-12514/com.fsck.k9 V/k9: conn42368062>>> 6 APPEND "Sent" (\Seen) {30411}
11-22 09:19:32.926 12496-12514/com.fsck.k9 V/k9: conn42368062<<<#+# [Ready for additional command text.]
11-22 09:19:33.170 12496-12514/com.fsck.k9 V/k9: conn42368062<<<#<br# [BAD, Command Error. 10]
```

**Comment by DrSegatron** [comment]
I'm wondering if my issues have the same cause; when mail is marked as read in one of the clients, after a little while I get a new "ghost" mail in my desktop client. There is no new mail, just the notification and signalling of a new mail.

**Comment by HenkDemper** [comment]
If there is anything I can do to reproduce, test, look at code (?) or whatever, please let me know. I could imagine that removing and adding the IMAP account with this behaviour might get rid of the issue (I have other IMAP accounts in K9 that don't have this problem), but I rather keep it in this state, so if needed I can reproduce & test. Downside is that Inbox sync now only works one way: from Desktop to K9, not reverse. Please advise...

**Comment by HenkDemper** [comment]
I exported my Account setting for this account from K9, then removed the account and re-imported the saved Account settings. That got rid of the problem, now K9 can sync Read status and delete messages in sync with my other mail clients for the same account. So I guess something got stuck/corrupted in a K9 cache/setting...
Anyway I'm happy that it works again and will notify when it happens again in the future. Should I close this item (even though it has been marked as 'bug') ?

**Comment by philipwhiuk** [comment]
No. There's some bad behaviour here we need to fix.

**Comment by cpieloth** [comment]
I can confirm this behaviour. I see it on my two IMAP accounts with two devices:
**Device 1:**
- Android 5.0.1
- k9 Mail 5.207

**Device 2:**
- Android 4.1.2
- k9 Mail 5.207

**Desktop:**
- Windows 10
- Thunderbird Mail Client

Unfortunatly I did not check/compare my logs with this ticket.

**Comment by cpieloth** [comment]
Maybe related to #1437

**Comment by cpieloth** [comment]
Maybe related to #1437, #1748, #1155, #1809

**Comment by bam80** [comment]
Any progress? The behavior still reproducible?

**Comment by HenkDemper** [comment]
For what it is worth: since I originally reported this issue I have not encountered it since. I have switched to Android 7.1 though if that could be of influence...

**Comment by cpieloth** [comment]
The IMAP sync of deleted messages is working again on my devices:

**Device 1:**
- Android 5.0.1
- k9 Mail **5.402**

**Device 2:**
- Android 4.1.2
- k9 Mail **5.402**

**Desktop:**
- Windows 10
- Thunderbird Mail Client

Only k9 Mail was updated to latest version: 5.402

**Comment by trevorpeacock** [comment]
I have been having similar problems appending mail to an Office365 Business (Hosted Exchange) IMAP server (I'm not using your project, rather using python-imaplib). Specific Mail messages fail with a similar error (a HTML tag followed by "BAD Command Error. 10"). It does affect specific messages only (only ones I've had trouble with are from a Yahoo Group). I've tried truncating the message, and it succeeds, but narrowing down the byte at which appending fails doesn't seem to relate to the tag noted in the error.
Sample log below.
Hopefully related and useful information, apologies otherwise. I've seen no documentation that helps me with this yet. Can provide samples of emails if that helps.

  08:12.47 > CNKF11 APPEND "Newsletters/FluteNet" () "12-Jan-2016 18:02:41 +0000" {39146}
  08:12.72 < + Ready for additional command text.
  08:12.72 write literal size 39146
  08:13.16 < <div BAD Command Error. 10

PS. I also saved the message and tried importing it to the same IMAP server via thunderbird. Couldn't get an error, the message was not imported.

**Comment by yang-dev** [comment]
not K-9 related but imaplib append keeps failing with `imaplib.abort: command: APPEND => unexpected response: b'}</style><!--[if BAD Command Error. 10' `when `}</style><!--[if` in message

**Comment by alexhenrie** [comment]
I have had a similar problem with [Geary](https://wiki.gnome.org/Apps/Geary), and after finding and reading both this bug report and https://github.com/jstedfast/MailKit/issues/542, I think the problem is that my message has a line that is nearly 30,000 characters long. Could your problem also be a line that is longer than the IMAP standard allows?

**Comment by cketti** [comment]
IMAP shouldn't care about the line length. However, RFC 5322 only allows 1000 characters per line.

K-9 Mail had a bug where it could have generated a message that exceeded this line length limit (see #822). This particular bug has been fixed in K-9 Mail 5.700.

Closing this issue because I don't think there's more we can learn from the logs provided.

**Comment by HenkDemper** [comment]
And to add from the original issuer of this ticket:

- I don't think (but can not tell for sure) that I normally have (or had) mails with ultra-long lines (1000 or even 30.000 chars), but then again can't be 100% sure
- I have been using K9 mail ever since and always auto-updated to the latest version. I have not experienced the raised problem since a long time, so I consider it fixed in 'some' (5.700 or other) version

Thanks for everybody contributing to K9, great work !

**Comment by techgeek-khushi-2029** [comment]
I recently came across an issue related to Thunderbird where the client throws a Command Error 12 while connecting to IMAP servers.
While researching it, I found this explanation that discusses the causes and possible fixes:
https://www.thunderbirdconverter.com/blog/thunderbird-command-error-12/

I'm wondering if errors like this could affect mailbox processing or data extraction workflows when working with large MBOX files.
